src/Features/Core/Portable/CodeRefactorings/SyncNamespace/AbstractSyncNamespaceService.ChangeNamespaceCodeAction.cs

```diff
@@ -0,0 +1,526 @@
+                    }
+
+                    var document = solution.GetDocument(id);
+                    LinkedDocumentsToSkip.AddRange(document.GetLinkedDocumentIds());
+
+                    document = await RemoveUnnecessaryImportsAsync(
+                        document, 
+                        CreateImports(document, names, withFormatterAnnotation: false), 
+                        cancellationToken)
+                        .ConfigureAwait(false);
+                    solution = document.Project.Solution;
+                }
+
+                LinkedDocumentsToSkip.Free();
+                return solution;
+            }
+
+            private async Task<Document> RemoveUnnecessaryImportsAsync(
+                Document document, 
+                IEnumerable<SyntaxNode> importsToRemove,
+                CancellationToken cancellationToken)
+            {
+                var removeImportService = document.GetLanguageService<IRemoveUnnecessaryImportsService>();
+                return await removeImportService.RemoveUnnecessaryImportsAsync(
+                    document,
+                    import => importsToRemove.Any(importToRemove => importToRemove.IsEquivalentTo(import, topLevel: false)),
+                    cancellationToken)
+                    .ConfigureAwait(false);
+            }
+
+            private async Task<Document> AddImportsInContainersAsync(
+                Document document,
+                IAddImportsService addImportService,
+                ImmutableArray<SyntaxNode> containers, 
+                ImmutableArray<string> names,
+                bool placeSystemNamespaceFirst, 
+                CancellationToken cancellationToken)
+            {
+                // Sort containers based on their span start, to make the result of 
+                // adding imports deterministic. This is mostly to make unit tests pass 
+                // and might not necessary for the correctness of the refactoring. 
+                // However, consider the number of containers will always be very small
+                // (if there's more than one at all), I decide to leave this in the product
+                // as it is very unlikely to affect perf.
+                if (containers.Length > 1)
+                {
+                    containers = containers.Sort(SyntaxNodeSpanStartComparer.Instance);
+                }
+
+                var imports = CreateImports(document, names, withFormatterAnnotation: true);
+                foreach (var container in containers)
+                {
+                    // If the container is a namespace declaration, the context we pass to 
+                    // AddImportService must be a child of the declaration, otherwise the 
+                    // import will be added to root node instead.
+                    var contextLocation = container is TNamespaceDeclarationSyntax 
+                        ? container.DescendantNodes().First() 
+                        : container;
```

Comment:
could we not just fix AddImportService?  Presumably it is walking to the parent of the container that is passed in.  But does it need to do that?  If it doesn't, would things still work for existing users of that service?

Reply:
I will follow up on this once this PR is done. Won't fix here

---
In reply to: [230477008](https://github.com/dotnet/roslyn/pull/30156#discussion_r230477008) [](ancestors = 230477008)